### PR TITLE
Add Notifications coloumn in alerts table

### PR DESCRIPTION
**Pull request by vikram-raj**
Story - https://issues.redhat.com/browse/ODC-4119

**Description**
This PR
- Adds the Notification column to the Alerts table and switch button to silence the notification for the alert.
- Adds the kebab action menu to view the alerting rule detail page.

Screenshots
Screenshot 2020-07-10 at 6 48 47 PM

Screenshot 2020-07-10 at 6 49 05 PM

Screenshot 2020-07-22 at 9 55 21 AM

Screenshot 2020-07-10 at 6 50 18 PM

Screenshot 2020-07-21 at 9 42 16 AM

### CODE REVIEW

**Comment by vikram-raj** [comment]
/hold
Depend on PR https://github.com/openshift/console/pull/5785

File: frontend/packages/dev-console/src/components/monitoring/alerts/SilenceDurationDropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import { coFetchJSON } from '@console/internal/co-fetch';
+import { Rule } from '@console/internal/components/monitoring/types';
+
+type SilenceDurationDropDownProps = {
+  rule: Rule;
+  namespace: string;
+};
+
+const SILENCE_FOR = 'Silence for ';
+const durations = [SILENCE_FOR, '30m', '1h', '2h', '6h', '12h', '1d', '2d', '1w'];
```
Comment (kyoto) [inline frontend/packages/dev-console/src/components/monitoring/alerts/SilenceDurationDropdown.tsx:19]: Mocks have dropdown labels like `30 minutes`, `1 hour`, etc.

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx [hunk 2]
```diff
@@ -58,15 +60,22 @@ export const monitoringAlertColumn: MonitoringAlertColumn[] = [
...
+  namespace: string,
+) => {
   const rows = [];
   _.forEach(alertrules, (rls) => {
     rows.push({
       ...(rls.state !== RuleStates.Inactive && {
-        isOpen: rls.state === RuleStates.Firing && !_.includes(collapsedRowsIds, rls.name),
+        isOpen:
+          (rls.state === RuleStates.Firing && !_.includes(collapsedRowsIds, rls.name)) ||
+          (rls.state !== RuleStates.Firing && _.includes(collapsedRowsIds, rls.name)),
```
Comment (kyoto) [inline frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:78]: Is this right? Shouldn't collapsed rows always be `isOpen = false`?

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx [hunk 3]
```diff
@@ -58,15 +60,22 @@ export const monitoringAlertColumn: MonitoringAlertColumn[] = [
...
+  collapsedRowsIds: string[],
+  namespace: string,
+) => {
   const rows = [];
   _.forEach(alertrules, (rls) => {
     rows.push({
       ...(rls.state !== RuleStates.Inactive && {
-        isOpen: rls.state === RuleStates.Firing && !_.includes(collapsedRowsIds, rls.name),
+        isOpen:
+          (rls.state === RuleStates.Firing && !_.includes(collapsedRowsIds, rls.name)) ||
```
Comment (kyoto) [inline frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:77]: Not introduced by this PR, but it now occurs to me that using the rule `name` as row key will fail if we have two rules with the same `name`.

**Comment by kyoto** [comment]
Noticed that the Notifications switch state is getting switched back to on when the poller fetches fresh data.

Also, the alignment is off at narrow screen widths:
screenshot

File: frontend/packages/dev-console/src/components/monitoring/alerts/SilenceAlert.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+  const handleChange = (checked: boolean) => {
+    setIsChecked(checked);
+  };
+
+  return (
+    <Switch
+      aria-label="Silence switch"
+      className="odc-silence-alert"
+      label="On"
```
Comment (kyoto) [inline frontend/packages/dev-console/src/components/monitoring/alerts/SilenceAlert.tsx:37]: I think you can just omit the `label` prop. Then you don't need the `.pf-c-switch__label.pf-m-on` `display: none` CSS.

Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/monitoring/alerts/SilenceAlert.tsx:37]: To add `labelOff`  prop `label` prop is required.

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx (hunk 3 above)
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:77]: Isn't `alertname` is going to be unique for a namespace?

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx (hunk 2 above)
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:78]: yes, collapsed rows always be `isOpen = false`. But when the user expands it should be expanded to handle this scenario I added this. I missed this scenario in my last PR.

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx (hunk 3 above)
Comment (kyoto) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx:79]: Ah, I forgot this was constrained by namespace.

So, this is the same situation as our silence rule discussion. `alertname` is not technically guaranteed to be unique even within a namespace, but in practice it almost certainly will be. I think this is acceptable for now, but we should probably have a more robust solution at some point.

File: frontend/public/components/monitoring/alerting.tsx [hunk 5]
```diff
@@ -179,14 +179,14 @@ const SilenceState = ({ silence }) => {
...
-const StateTimestamp = ({ text, timestamp }) => (
+export const StateTimestamp = ({ text, timestamp }) => (
   <div className="text-muted monitoring-timestamp">
     {text}&nbsp;
     <Timestamp timestamp={timestamp} />
   </div>
 );
 
-const AlertStateDescription = ({ alert }) => {
+export const AlertStateDescription = ({ alert }) => {
```
Comment (kyoto) [inline frontend/public/components/monitoring/alerting.tsx:189]: Looks like we don't need to export this

File: frontend/public/components/monitoring/alerting.tsx [hunk 6]
```diff
@@ -481,7 +481,7 @@ const AlertMessage: React.FC<AlertMessageProps> = ({ alertText, labels, template
   return <p>{messageParts}</p>;
 };
 
-const HeaderAlertMessage: React.FC<{ alert: Alert; rule: Rule }> = ({ alert, rule }) => {
+export const HeaderAlertMessage: React.FC<{ alert: Alert; rule: Rule }> = ({ alert, rule }) => {
```
Comment (kyoto) [inline frontend/public/components/monitoring/alerting.tsx:484]: Looks like we don't need to export this

File: frontend/packages/dev-console/src/components/monitoring/alerts/SilenceDurationDropdown.tsx (hunk 1 above)
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/monitoring/alerts/SilenceDurationDropdown.tsx]: Done

File: frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx (hunk 3 above)
Comment (vikram-raj) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx]: Fixed it, now I am using rules id instead of the name.

Comment (kyoto) [reply frontend/packages/dev-console/src/components/monitoring/alerts/monitoring-alerts-utils.tsx]: 👍

File: frontend/packages/dev-console/src/components/monitoring/alerts/SilenceAlert.tsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+type SilenceAlertProps = {
+  rule: Rule;
+  namespace: string;
+};
+
+const { alertManagerBaseURL } = window.SERVER_FLAGS;
+
+const SilenceUntil = ({ rule }) => {
+  if (alert && !_.isEmpty(rule.silencedBy)) {
```
Comment (kyoto) [inline frontend/packages/dev-console/src/components/monitoring/alerts/SilenceAlert.tsx:19]: This `alert &&` looks unnecessary

File: frontend/packages/dev-console/src/components/monitoring/alerts/SilenceAlert.tsx (hunk 4 above)
Comment (kyoto) [reply frontend/packages/dev-console/src/components/monitoring/alerts/SilenceAlert.tsx:45]: Looks like you can set `label` to `null`, then you can still use `labelOff` without needing the CSS.

File: frontend/packages/dev-console/src/components/monitoring/alerts/SilenceAlert.tsx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    return <StateTimestamp text="Until" timestamp={_.max(_.map(rule.silencedBy, 'endsAt'))} />;
+  }
+  return null;
+};
+
+const SilenceAlert: React.FC<SilenceAlertProps> = ({ rule, namespace }) => {
+  const [isChecked, setIsChecked] = React.useState(true);
+  React.useEffect(
+    () => (rule.state === RuleStates.Silenced ? setIsChecked(false) : setIsChecked(true)),
```
Comment (kyoto) [inline frontend/packages/dev-console/src/components/monitoring/alerts/SilenceAlert.tsx:28]:
```suggestion
    () => setIsChecked(rule.state !== RuleStates.Silenced)
```

File: frontend/packages/dev-console/src/components/monitoring/alerts/SilenceAlert.tsx [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+type SilenceAlertProps = {
+  rule: Rule;
+  namespace: string;
+};
+
+const { alertManagerBaseURL } = window.SERVER_FLAGS;
+
+const SilenceUntil = ({ rule }) => {
+  if (alert && !_.isEmpty(rule.silencedBy)) {
+    return <StateTimestamp text="Until" timestamp={_.max(_.map(rule.silencedBy, 'endsAt'))} />;
```
Comment (kyoto) [inline frontend/packages/dev-console/src/components/monitoring/alerts/SilenceAlert.tsx:20]:
```suggestion
    return <StateTimestamp text="Until" timestamp={_.maxBy(rule.silencedBy, 'endsAt')} />;
```

File: frontend/packages/dev-console/src/components/monitoring/alerts/SilenceDurationDropdown.tsx [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  '2h': '2 hours',
+  '1d': '1 day',
+};
+
+const { alertManagerBaseURL } = window.SERVER_FLAGS;
+
+const SilenceDurationDropDown: React.FC<SilenceDurationDropDownProps> = ({ rule, namespace }) => {
+  const createdBy = useSelector((state: RootState) => state.UI.get('user')?.metadata?.name);
+
+  const matchers = [
```
Comment (kyoto) [inline frontend/packages/dev-console/src/components/monitoring/alerts/SilenceDurationDropdown.tsx:30]: Should also add matchers for any additional labels found in `rule.labels`.

File: frontend/public/reducers/ui.ts [hunk 11]
```diff
@@ -59,6 +59,26 @@ const silenceFiringAlerts = (firingAlerts, silences) => {
   });
 };
 
+const silenceFiringRules = (firingRules, silences) => {
```
Comment (kyoto) [inline frontend/public/reducers/ui.ts:62]: Looks like we can make this much simper by removing `silenceFiringRules` and leaving `MonitoringSetRules` unchanged. Then modify `silenceFiringAlerts` instead.

`silenceFiringAlerts` is already being run when the rules are imported, and already does much of the work to determine which alerts are silenced.

Modify `silenceFiringAlerts ` by adding the following (or something similar) inside the `if (a.silencedBy.length) { ... }` block.
```
if (!_.isEmpty(a.rule.alerts) && _.every(a.rule.alerts, isSilenced)) {
  a.rule.state = RuleStates.Silenced;
  a.rule.silencedBy = _.filter(
    silences?.data,
    (s) => s.status.state === SilenceStates.Active && _.some(a.rule.alerts, isSilenced),
  );
}
```
(briefly tested, but not properly)

**Pull request merged**